### PR TITLE
Juju 7122/add methods for role management

### PR DESCRIPTION
**Pull request by ale8k**
## Description

Adds database methods for roles according to the specification. 

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [x] Covered by integration tests

### CODE REVIEW

File: internal/db/role.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	if err := d.DB.WithContext(ctx).Delete(role).Error; err != nil {
+		return errors.E(op, dbError(err))
+	}
+	return nil
+}
+
+// ForEachRole iterates through all role entries applying the provided callback function.
+func (d *Database) ForEachRole(ctx context.Context, f func(*dbmodel.RoleEntry) error) (err error) {
```
Comment (SimoneDutto) [inline internal/db/role.go]: group has ListGroup, which is more suited for our listing use cases. So either replace this forEachRole with ListRoles, or add ListRoles

File: internal/db/role_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		Name: "test-role",
+	}
+	tx := s.Database.DB.First(&re)
+	c.Assert(tx.Error, qt.IsNil)
+	c.Assert(re.ID, qt.Equals, uint(1))
+	c.Assert(re.Name, qt.Equals, "test-role")
+	c.Assert(re.UUID, qt.Equals, uuid)
+}
+
+func (s *dbSuite) TestDatabase_GetRole(c *qt.C) {
```
Comment (SimoneDutto) [inline internal/db/role_test.go]: i see you want to bring the underscore in this Go Land

File: internal/db/role_test.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	ge1.Name = "renamed-role"
+	err = s.Database.UpdateRole(context.Background(), ge1)
+	c.Check(err, qt.IsNil)
+
+	ge2 := &dbmodel.RoleEntry{
+		Name: "renamed-role",
+	}
+	err = s.Database.GetRole(context.Background(), ge2)
+	c.Check(err, qt.IsNil)
+	c.Assert(ge2, qt.DeepEquals, ge1)
```
Comment (SimoneDutto) [inline internal/db/role_test.go]: how they are deep equals? At least their `UpdatedAt` should be different, right?

File: internal/db/role.go
Comment (SimoneDutto) [inline internal/db/role.go:1]: we need the same methods implemented for the groups.
For example, here we are missing `CountRoles` and this is needed for rebac admin pagination

**Review by SimoneDutto — CHANGES_REQUESTED** [review]

File: internal/db/role.go (hunk 1 above)
Comment (ale8k) [reply internal/db/role.go]: The spec didn't say ListRoles, but I agree @alesstimec any reason ForEachRole and not ListRoles?

File: internal/db/role_test.go (hunk 2 above)
Comment (ale8k) [reply internal/db/role_test.go]: Aha I do ;)

File: internal/db/role_test.go (hunk 3 above)
Comment (ale8k) [reply internal/db/role_test.go]: Fair point! I'm not actually sure...

**Review by SimoneDutto — APPROVED** [review]
lgtm

File: internal/db/role.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	if err := d.ready(); err != nil {
+		return errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	db := d.DB.WithContext(ctx)
+	if role.ID != 0 {
```
Comment (alesstimec) [inline internal/db/role.go:51]: should we have some validation to make sure at least one of `id`, `uuid` or `name` is non-empty?

File: internal/db/role.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		db = db.Where("name = ?", role.Name)
+	}
+	if err := db.First(&role).Error; err != nil {
+		return errors.E(op, dbError(err))
+	}
+	return nil
+}
+
+// UpdateRole updates the role identified by its ID or UUID.
+func (d *Database) UpdateRole(ctx context.Context, role *dbmodel.RoleEntry) (err error) {
```
Comment (alesstimec) [inline internal/db/role.go]: suggestion: UpdateRoleName

File: internal/db/role.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	const op = errors.Op("db.UpdateRole")
+
+	if role.ID == 0 {
+		return errors.E(errors.CodeNotFound)
+	}
+	if role.UUID == "" {
+		return errors.E("role uuid not specified", errors.CodeNotFound)
+	}
```
Comment (alesstimec) [inline internal/db/role.go:72]: code above contradicts the godoc: both `ID` and `UUID` must be specified. Is this what we want?

File: internal/db/role.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+// RemoveRole removes the role identified by its ID or UUID.
+func (d *Database) RemoveRole(ctx context.Context, role *dbmodel.RoleEntry) (err error) {
+	const op = errors.Op("db.RemoveRole")
+
+	if role.ID == 0 {
+		return errors.E(errors.CodeNotFound)
+	}
+	if role.UUID == "" {
+		return errors.E(errors.CodeNotFound)
+	}
```
Comment (alesstimec) [inline internal/db/role.go:97]: same here: code requires that both be specified.

File: internal/db/role_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+		Name: "test-role",
+	}
+	tx := s.Database.DB.First(&re)
+	c.Assert(tx.Error, qt.IsNil)
+	c.Assert(re.ID, qt.Equals, uint(1))
+	c.Assert(re.Name, qt.Equals, "test-role")
+	c.Assert(re.UUID, qt.Equals, uuid)
+}
+
+func (s *dbSuite) TestDatabase_GetRole(c *qt.C) {
```
Comment (alesstimec) [inline internal/db/role_test.go]: veto: please no underscores.

File: internal/db/role_test.go (hunk 3 above)
Comment (alesstimec) [reply internal/db/role_test.go]: worth investigating.

**Review by alesstimec — CHANGES_REQUESTED** [review]

File: internal/db/role_test.go (hunk 3 above)
Comment (SimoneDutto) [reply internal/db/role_test.go]: I think gorm is not aware of UpdateAt. 

When we are not using gorm in the models maybe we've also lost this autoupdating mechanism for `CreatedAt, UpdatedAt`. Please investigate this or create a card

File: internal/db/role.go (hunk 4 above)
Comment (ale8k) [reply internal/db/role.go:51]: Is there a need? We have the not-found error, I imagine this is good enough?

File: internal/db/role.go (hunk 5 above)
Comment (ale8k) [reply internal/db/role.go]: I'll update groups too (so they look similar)

File: internal/db/role.go (hunk 6 above)
Comment (ale8k) [reply internal/db/role.go:72]: I think so, you'll get the role and it'll have an id and uuid. I guess it's an "extra safe" check.

File: internal/db/role.go (hunk 5 above)
Comment (SimoneDutto) [reply internal/db/role.go]: wait, but this method is implicitly used only to update the name. But you can potentially update any field, right? So i think the name is correct.

File: internal/db/role_test.go (hunk 3 above)
Comment (ale8k) [reply internal/db/role_test.go]: Looking now

File: internal/db/role.go (hunk 4 above)
Comment (kian99) [reply internal/db/role.go:51]: I suggest adding a test with no ID, name or UUID specified to verify that behavior.

File: internal/db/role.go (hunk 6 above)
Comment (kian99) [reply internal/db/role.go:72]: We should update the godoc in that case.

**Review by kian99 — COMMENTED** [review]
Just gave it a quick review so won't approve but looks good.

File: internal/db/role.go (hunk 4 above)
Comment (ale8k) [reply internal/db/role.go:51]: I modified the existing with name set to have nothing.

File: internal/db/role.go (hunk 6 above)
Comment (ale8k) [reply internal/db/role.go:72]: Done

File: internal/db/role_test.go (hunk 3 above)
Comment (ale8k) [reply internal/db/role_test.go]: Created bug.

File: internal/db/role.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+// UpdateRoleName updates the role name identified by its ID and UUID.
+func (d *Database) UpdateRoleName(ctx context.Context, role *dbmodel.RoleEntry) (err error) {
+	const op = errors.Op("db.UpdateRole")
+
+	if role.ID == 0 {
+		return errors.E(errors.CodeNotFound)
+	}
+	if role.UUID == "" {
+		return errors.E("role uuid not specified", errors.CodeNotFound)
```
Comment (alesstimec) [inline internal/db/role.go]: in should be able to do 
```
UpdateRoleName(ctx, &dbmodel.RoleEntry{
  ID: <id>,
  Name: "some-new-name",
}
``` 
or 
```
UpdateRoleName(ctx, &dbmodel.RoleEntry{
  UUID: <uuid>,
  Name: "some-new-name",
}
```
we don't need both to be filled in to update the role name

File: internal/db/role.go (hunk 7 above)
Comment (alesstimec) [reply internal/db/role.go:97]: one i enough to uniquely identify the role we want to remove

File: internal/db/role_test.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+	err := s.Database.GetRole(context.Background(), &dbmodel.RoleEntry{})
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeUpgradeInProgress)
+
+	err = s.Database.Migrate(context.Background(), false)
+	c.Assert(err, qt.IsNil)
+
+	role := &dbmodel.RoleEntry{}
+	err = s.Database.GetRole(context.Background(), role)
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeNotFound)
```
Comment (alesstimec) [inline internal/db/role_test.go]: hmm.. but if you add a role above this, then we expect to get one role back - whichever is first by some magic order.

File: internal/db/role.go (hunk 4 above)
Comment (alesstimec) [reply internal/db/role.go:51]: I would still require at least one of these fields that (each) uniquely identify a role to be set

**Review by alesstimec — CHANGES_REQUESTED** [review]

File: internal/db/role.go (hunk 9 above)
Comment (ale8k) [reply internal/db/role.go]: Done, one must be set now.

File: internal/db/role_test.go (hunk 10 above)
Comment (ale8k) [reply internal/db/role_test.go]: Would you like me to break it into its own test? If I do, it'd probably be worth breaking up all cases.  But many of our tests bundle many cases into one so it'd be a bit of an outlier

File: internal/db/role.go (hunk 7 above)
Comment (ale8k) [reply internal/db/role.go:97]: Updated to be either or

File: internal/db/role.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	model := d.DB.WithContext(ctx).Model(&dbmodel.RoleEntry{})
+	model.Where("uuid = ?", uuid)
+	if model.Update("name", name).RowsAffected == 0 {
+		return errors.E(op, "role does not exist")
```
Comment (alesstimec) [inline internal/db/role.go]:
```suggestion
		return errors.E(op, "role not found", errors.CodeNotFound)
```

**Review by alesstimec — APPROVED** [review]
LGTM with one suggestion

**Pull request merged**